Inform about silent supergroup migrations (#559)
grammyjs/website: site/docs/plugins/session.md

Make sure you understand the consequences of this configuration by reading [this](../guide/deployment-types.md) article and especially [this](./runner.md#sequential-processing-where-necessary) one.
:::

### Supergroup Migrations

If you are using sessions for a group chat you should be aware that telegram migrates regular groups to supergroups under certain circumstances (e.g. [here](https://github.com/telegramdesktop/tdesktop/issues/5593)). 

This migration can only occur once per group chat but it can cause inconsistencies. This is because the migrated chat is technically a completely different chat with a different `chat.id` and hence a different session id.

Currently there is no safe solution to this problem because the two chats can send messages simultaneously leading to data races. However, there are several ways of dealing with this issue:

* Ignoring the problem. The bot's session data will effectively reset when a group is migrated. Simple, reliable, default behavior, but potentially unexpected once per chat. For example if a migration happens while a user is in a conversation (the conversation plugin uses sessions) the conversation will be reset.

> * Only storing temporary data (or data with timeouts) in the session, and using a database for the important things that need to be migrated when a chat migrates. This can then use transactions and custom logic to handle concurrent data access from the old and the new chat. A lot of effort and has a performance cost, but the only truly reliable way to solve this problem.

Reviewer: @roj1512 what's your opinion on this?

> * Only storing temporary data (or data with timeouts) in the session, and using a database for the important things that need to be migrated when a chat migrates. This can then use transactions and custom logic to handle concurrent data access from the old and the new chat. A lot of effort and has a performance cost, but the only truly reliable way to solve this problem.

Author: I added two sentences in [#61b72f1](https://github.com/grammyjs/website/pull/559/commits/61b72f14fdbec4f4a20be58606056a9e123f8974). Maybe this puts it into context a bit?

> * Only storing temporary data (or data with timeouts) in the session, and using a database for the important things that need to be migrated when a chat migrates. This can then use transactions and custom logic to handle concurrent data access from the old and the new chat. A lot of effort and has a performance cost, but the only truly reliable way to solve this problem.

Reviewer: Looking good, thanks

Make sure you understand the consequences of this configuration by reading [this](../guide/deployment-types.md) article and especially [this](./runner.md#sequential-processing-where-necessary) one.
:::

### Chat Migrations

> If you are using sessions for groups, you should be aware that Telegram migrates regular groups to supergroups under certain circumstances (e.g. [here](https://github.com/telegramdesktop/tdesktop/issues/5593)).

Reviewer 2: How is this Telegram Desktop issue related?

Author: It's just the only example / documentation for the impact of the migration I could find online. It's just an example of how to distinguish supergroups from normal groups (the part about "Actual behaviour"). Nothing in particular about Telegram Desktop.

Reviewer 2: Still a little misleading, don’t you think?

Author: Well, I don't know. Tbh I did not even notice that this is about Telegram Desktop. We could also use [this link](https://stackoverflow.com/questions/54555821/why-telegram-group-automatically-becomes-super-group) instead

Reviewer 2: Can you remove the link?

Reviewer: It's documented at https://core.telegram.org/bots/api#message, too. Given how this is a main document that grammY is based on, it could make sense to change the link to that. Given how we constantly repeat information that is found on this page without naming it again, we might just as well remove the link. You decide :)

Author: Okay, then I'll just remove the link and leave it to the reader :P 
Don't have access to my computer right now, will update it at the end of the week 👍

Reviewer 2: @Madnex I think you can just rest in your weekend. Thanks for the suggestion, @KnorpelSenf.

Reviewer: Awesome, thanks Roj. I can offer to review.

> If you are using sessions for a group chat you should be aware that Telegram migrates regular groups to supergroups under certain circumstances (e.g. [here](https://github.com/telegramdesktop/tdesktop/issues/5593)).

Reviewer 2: ```suggestion
If you are using sessions for groups, you should be aware that Telegram migrates regular groups to supergroups under certain circumstances (e.g. [here](https://github.com/telegramdesktop/tdesktop/issues/5593)).
```

> This migration can only occur once per group chat but it can cause inconsistencies.

Reviewer 2: ```suggestion
This migration only occurs once for each group, but it can cause inconsistencies.
```

> This is because the migrated chat is technically a completely different chat with a different chat identifier and hence a different session identifier.

Reviewer 2: ```suggestion
This is because the migrated chat is technically a completely different chat that has a different chat identifier, and hence a different session identifier.
```

> Currently, there is no safe solution to this problem because the two chats can send messages simultaneously.

Reviewer 2: ```suggestion
Currently, there is no safe solution for this problem because messages from the two chats are also differently identified.
```

This can lead to data races.
However, there are several ways of dealing with this issue:

- Ignoring the problem.
  The bot's session data will effectively reset when a group is migrated.
  Simple, reliable, default behavior, but potentially unexpected once per chat.

>   For example if a migration happens while a user is in a conversation (the conversation plugin uses sessions) the conversation will be reset.

Reviewer 2: ```suggestion
  For example, if a migration happens while a user is in a conversation powered by the [conversations plugin](./conversations.md), the conversation will be reset.
```

- Only storing temporary data (or data with timeouts) in the session, and using a database for the important things that need to be migrated when a chat migrates.
  This can then use transactions and custom logic to handle concurrent data access from the old and the new chat.
  A lot of effort and has a performance cost, but the only truly reliable way to solve this problem.

- It is theoretically possible to implement a workaround that matches both chats **without guarantee of reliability**.
  The Telegram Bot API sends a migration update for each of the two chats once the migration was triggered (see the properties `migrate_to_chat_id` or `migrate_from_chat_id` in the [Telegram API Docs](https://core.telegram.org/bots/api#message)).
  The issue is that there is no guarantee that these messages are sent before a new message in the supergroup appears.
  Hence, the bot could receive a message from the new supergroup before it is aware of any migration and thus, it can not match the two chats, resulting in the aforementioned problems.

> - Another workaround would be to limit the bot only for supergroups with filtering (or limit only session related features to supergroups).

Reviewer 3: ```suggestion
- Another workaround would be to limit the bot only for supergroups with [filtering](../guide/filter-queries.md) (or limit only session related features to supergroups).
```
